PEP 605: A rolling feature release stream for CPython (#1184)
python/peps: pep-0605.rst

PEP: 605
Title: A rolling feature release stream for CPython
Version: $Revision$
Last-Modified: $Date$
Author: Steve Dower <steve.dower@python.org>, Nick Coghlan <ncoghlan@gmail.com>
Discussions-To: Link TBD
Status: Draft
Type: Informational
Content-Type: text/x-rst
Created: 20-Sep-2019
Python-Version: 3.9
Post-History: 30-Sep-2019

Abstract
========

> Rather than proposing a change to CPython's release cadence (as PEP 602 does),

Author: Adopted a similar-in-spirit rewording.

or a policy change to allow backwards compatible feature additions later in a
release series (as PEP 598 does), this PEP instead proposes that we create a
rolling stream of production-ready beta releases by amending CPython's
pre-release management process as follows:

* Feature freeze, ABI freeze, pyc file format freeze, and maintenance branch
  creation all correspond with the first release candidate for an X.Y.0 release
* The X.Y.0 release candidate period is extended from 3 weeks to 2 months
* Beta releases from the master git branch would occur every 2 months
  whenever there is no X.Y.0 release in the release candidate phase
* Routine alpha releases are removed from the process
* Release managers may still choose to mark particular releases as alpha
  releases if they feel such releases are appropriate. If any such alpha
  releases are made, it is expected that they would be in place of one of the
  scheduled beta releases

> * The deprecation policy for interfaces that have been included in an X.Y.0

Reviewer: PEP 387 should handle the deprecation policy specification.

Author: Done

  release or release candidate without being explicitly declared provisional
  remains unchanged (incompatible changes are preferably avoided in the first
  place, but if they do take place, should emit ``DeprecationWarning`` for at

>   least one full preceding X.Y.Z release series)

Reviewer: I take "X.Y.Z release series" to mean "for the lifetime of X.Y", correct?

* Each new alpha or beta release may introduce new deprecation warnings for

>   interfaces that appeared in the previous X.Y.Z release series

Author: Agreed, done.

* Each new alpha or beta release may change or remove interfaces that either did
  not appear in the previous X.Y.Z release series, or emitted a relevant
  deprecation warning in that series

> * Additions to the stable C ABI are permanent once they have appeared in at

Reviewer: So there's absolutely no deprecation of any part of the stable ABI? Even going from X to X+1 major versions?

Author: There's a version number in the stable ABI (`abi3`). The implication is that we'd change it to `abi4` in Python 4.0, but there's no actual obligation to do so.

  least one beta release
* All other newly added interfaces that have appeared only in alpha or beta
  releases are considered provisional, and may be changed in incompatible ways
  without requiring a preceding deprecation warning

As part of this change, the current "not for production use" guidance given for
beta releases would be amended to state "suitable for production use only in
environments with sufficiently robust compatibility testing and operational
monitoring capabilities".

> This PEP also proposes that the frequency of X.Y.0 releases be adjusted to

Reviewer: But the opening of this PEP said "Rather than proposing a change to CPython's release cadence" which is clearly is a shift of.

Author: Yep, reworded

begin each new release series in August every two years (starting in 2021,
around two years after the release of Python 3.8.0).

Motivation
==========

The current CPython pre-release and release management processes were developed
in an era where automated continuous integration and operational monitoring
systems were still relatively immature. Since that time, many organisations
have adopted deployment models that allow them to incorporate new CPython
feature releases without adding substantially more risk than they incur for any

> other code change.

Author: Added Yet Another Sentence :)

In addition to allowing more time for end user feedback on the release
candidate, this adjusted policy also provides additional time for maintainers
of Python projects to build and publish pre-built wheel archives for the new
stable release series, significantly improving the initial user experience of
the X.Y.0 release.

Removal of routine alpha releases
---------------------------------

Due to the proposed policy changes for beta releases and release candidates,
the traditional CPython alpha release period would no longer serve a clear
purpose, and as such would be skipped, with the first pre-release in a new
series being X.Y.0b1.

However, when it seems appropriate to do so, release managers may choose to
declare that a scheduled beta release will be marked as an alpha release
instead. For example, a release manager may choose to do this when major changes
have been made to the release management tooling, when a particularly
large feature has landed that the release manager views as having a high chance
of introducing unintended side effects, or when a new interface has been added
to the stable C ABI that requires further testing and design feedback before

> being locked down in a beta release.

Reviewer 2: Is it obvious enough that the stable C ABI "restriction" (permanent as soon as published in beta) can be avoided by just not putting the new APIs into the stable ABI on first release? That seems obvious to me, but perhaps we should call it out explicitly?

Reviewer: I would call it out explicitly. I thought it was a bit extreme until I realized that it could go into the CPython API first.

Author: The next push adds a parenthetical example that tries to make this clearer (there are some functions that only have purpose if they're part of the stable ABI, so they can't go through the usual CPython ABI -> stable ABI promotion path).

which preserves the status quo in that respect.

Caveats and Limitations
=======================

Actual release dates may be scheduled up to a month earlier or later at
the discretion of the release manager, based on release team availability, and
the timing of other events (e.g. PyCon US, or the annual core developer
sprints). However, as one goal of the proposal is to provide a consistent
release cadence, adjustments should ideally be rare.

Within a release series, the exact frequency of maintenance releases would
still be up to the release manager and the binary release team; this PEP
only proposes an expected cadence for pre-releases and X.Y.0 releases.

However, for the sake of the example timelines, the PEP assumes maintenance
releases every other month, allowing them to alternate months with the rolling
beta releases.

Design Discussion
=================

Why rolling beta releases over simply doing more frequent X.Y.0 releases?
-------------------------------------------------------------------------

For large parts of Python's user base, *availability* of new CPython feature
releases isn't the limiting factor on their adoption of those new releases

> (this effect shows through clearly in such metrics as PyPI download metadata).

Reviewer 2: ```suggestion
(this effect is visible in such metrics as PyPI download metadata).
```

As such, any proposal based on speeding up full feature releases needs to strike
a balance between meeting the needs of users who would be adopting each release
as it became available, and those that would now be in a position of adopting
every 2nd, 3rd, or 4th release, rather than being able to migrate to almost
every release at some point within its lifecycle.

This proposal aims to approach the problem from a different angle by defining a
*new* production-ready release stream that is more specifically tailored to the

> interests of operating environments that are likely to be able to consume new

Reviewer 2: ```suggestion
interests of operating environments that are able to consume new
```

even if that feature flag was subsequently enabled by default before its first
appearance in an X.Y.0 release candidate.

The rationale behind these approaches is that explicit detection/enabling like
this would make it straightforward for users of the rolling beta release stream
to notice when we remove or change provisional features
(e.g. ``from __future__`` imports break on compile if the feature flag no
longer exists), or to safely fall back on previous functionality.

The interpreter's rich attribute lookup machinery means we can also choose to
add warnings for provisional or deprecated imports and attributes that we don't
have any practical way to add for checks against the value of
``sys.version_info``.

Implications for CPython core development
-----------------------------------------

The major change for CPython core development is the need to keep the master
branch more consistently release ready.

While the main requirement for that would be to keep the stable BuildBot fleet
green, there would also be encouragement to keep the development version of
the documentation up to date for the benefit of users of the rolling beta

> releases.

Reviewer: And specifically keeping "What's New" up-to-date which we currently do not do.

Author: Added a sentence for that.

> On the specific topic of the stable ABI, the "all ones" ``PY_VERSION_HEX``

Reviewer 2: Eh, I'd remove this section and just say that if we're not sure the API is stable, don't mark it as stable yet :)

Reviewer 3: Agreed, I think this section might be adding additional complexity to differentiating stable vs questionably stable/unstable when it could be done more simply. Maybe there's a need for it that I'm not seeing though.

Author: Done!

providers used solely in an advisory capacity for early detection of potential
compatibility problems.

Having a rolling beta release stream available may also make it more feasible
for more CI providers to offer a "CPython beta release" testing option. At the
moment, this feature is only available from CI providers that are willing and
able to put the necessary time and effort into creating, testing, and publishing
their own builds from the CPython master branch (e.g. [6_]).

Implications for the proposed Scientific Python ecosystem support period
------------------------------------------------------------------------

Based on discussions at SciPy 2019, NEP (NumPy Enhancement Proposal) 29 has
been drafted [3_] to propose a common convention across the Scientific Python
ecosystem for dropping support for older Python versions.

While the exact formulation of that policy is still being discussed, the initial
proposal is very simple: support any Python feature release published within
the last 42 months.

For an 18 month feature release cadence, that works out to always supporting at
least the two most recent feature releases, and then dropping support for all

> X.Y.z releases around 6 months after X.(Y+2).0 is released. This means there is

Reviewer: ```suggestion
X.Y.Z releases around 6 months after X.(Y+2).0 is released. This means there is
```

Reviewer 2: For Nick's benefit, I posted an alternative suggestion for this part below.

Author: Just went with capitalisation change for now, albeit for reasons that are probably to subtle to be worth caring about...

> X.Y.z releases around 6 months after X.(Y+2).0 is released. This means there is

Reviewer 2: `X.Y.z` is yet another notation for release series.

Perhaps "support for the X.Y release series around 6 months afte X.(Y+2).0 is released"?

Author: The wording here echoes some of the phrasing in NEP 29. The gist is that if X.Y.0 was released within the 42 month window, projects should aim to support all X.Y.Z releases, not just the most recently published one.

So this section is the only place where I've kept the 3-part formulation (I did fix the capitalisation, though, as Brett suggested).

a 6 month period roughly every other year where the three most recent feature
releases are supported.

For a 12 month release cadence, it would work out to always supporting at
least the three most recent feature releases, and then dropping support for all
X.Y.z releases around 6 months after X.(Y+3).0 is released. This means that
for half of each year, the four most recent feature releases would be supported.

For a 24 month release cadence, a 42 month support cycle works out to always
supporting at least the most recent feature release, and then dropping support

> for all X.Y.z feature releases around 18 months after X.(Y+1).0 is released.

Reviewer: ```suggestion
for all X.Y.Z feature releases around 18 months after X.(Y+1).0 is released.
```
